driver/amc_pci_core.c

```diff
@@ -125,19 +153,35 @@ static int amc_pci_open(struct inode *inode, struct file *file)
     /* Replace the file's f_ops with our own and perform any device specific
      * initialisation. */
     int minor_index = iminor(inode) - amc_priv->minor;
-    file->f_op = fops_info[minor_index].fops;
+
     int rc = -EINVAL;
-    switch (minor_index)
+    struct prom_entry *pentry = prom_find_entry(amc_priv->prom, minor_index);
+
+    if (pentry)
     {
-        case MINOR_REG:
-            rc = amc_pci_reg_open(file, amc_priv->dev, amc_priv->interrupts, &amc_priv->locking);
-            break;
-        case MINOR_DDR0:
-            rc = amc_pci_dma_open(file, amc_priv->dma, DDR0_BASE, DDR0_LENGTH);
-            break;
-        case MINOR_DDR1:
-            rc = amc_pci_dma_open(file, amc_priv->dma, DDR1_BASE, DDR1_LENGTH);
-            break;
+        if (pentry->tag == PROM_DEVICE_TAG)
```

Comment:
Think a switch statement on `pentry->tag` would be nice.  I also note we silently ignore unrecognised tags; is that the right behaviour?

Reply:
Done, although there is a little detail to consider, in C99, we get an error if a declaration happens immediately after a tag (which is used for each case), so I added a dummy ; statement to bypass that

Reply:
Ah.  Don't do declarations inside switch statements without an enclosing `{ ... }` block; the results are otherwise really rather ugly.

Reply:
fixed